File: frontend/src/services/sessionInput.ts
import { apiClient } from './api'
import type { QueryOptionResult } from './probeQuery'
import type { AiUsage } from './aiUsage'
import type { FinalQueryVariant } from './finalQuery'

export interface SessionInputRootPayload {
  theme: string
  description: string | null
  area_of_study: string | null
  keywords: string[] | null
  year_from: number | null
  year_to: number | null
}

export interface SessionInputGeneratedPayload {
  theme: string
  description: string | null
  iterations: number
}

interface FormInput {
  theme: string
  description: string | null
  keywords: string | null
  studyArea: string | null
}

// Converte o shape do useFormStore.input para o payload raiz esperado pelo
// endpoint /session-input, fazendo o split de keywords (string separada por
// vírgula) em uma lista sem vazios.
export function mapInputToSessionInputRoot(input: FormInput): SessionInputRootPayload {
  const keywordsArray = input.keywords
    ? input.keywords.split(',').map((k) => k.trim()).filter(Boolean)
    : null

  return {
    theme: input.theme,
    description: input.description || null,
    area_of_study: input.studyArea || null,
    keywords: keywordsArray && keywordsArray.length > 0 ? keywordsArray : null,
    year_from: null,
    year_to: null,
  }
}

export interface SessionProbeQueryPayload {
  fonte: 'ops' | 'scopus'
  // null = a query da probe (Exploração Inicial); specific/balanced/generic
  // = a variante de query final escolhida (Escolha da Query Final) -
  // auto-relacionada à linha de probe da mesma fonte (ver
  // db/research_session_models.py:SessionProbeQuery).
  tipo: FinalQueryVariant | null
  query_text: string
  fields: Record<string, string[]> | null
  year_from: number | null
  year_to: number | null
  complexity_score: number | null
  complexity_level: string | null
  iterations: number
  result_count: number | null
  // Dicts crus dos documentos encontrados por essa query (só um dos dois
  // preenchido, conforme `fonte`) - o backend faz o mapeamento campo-a-campo
  // e persiste em patent/article (ver session_probe_documents.py).
  patents: Record<string, unknown>[]
  articles: Record<string, unknown>[]
}

// Monta o payload de uma linha de session_probe_query pronta pra enviar no
// finalize/save - usado tanto pra linha de probe (tipo=null, Step3) quanto
// pra linha de query final (tipo=variante escolhida, Escolha da Query
// Final), a partir da opção selecionada numa seção. `iterations` é o
// contador interno da store (0-based: 0 = gerada uma vez, sem retry) - aqui
// vira o valor de negócio "nº de vezes que a query foi gerada" (1+); pra
// query final, o chamador já soma iterações de amostragem de termos +
// regeneração da query final antes de passar aqui. `resultCount` vem da
// busca real já rodada (Step3 ou Escolha da Query Final), null se ainda não
// rodou. Retorna null se não há uma seleção válida (nada gerado, ou a
// tentativa selecionada falhou) - mesma condição usada no `canProceed` do
// Step3/FinalExploration.
export function buildProbeQueryPayload(
  selected: QueryOptionResult | undefined,
  fonte: 'ops' | 'scopus',
  iterations: number,
  resultCount: number | null = null,
  rawItems: Record<string, unknown>[] = [],
  tipo: FinalQueryVariant | null = null,
): SessionProbeQueryPayload | null {
  if (!selected?.success) return null
  return {
    fonte,
    tipo,
    query_text: selected.query?.query ?? '',
    fields: selected.fields ?? null,
    year_from: selected.year_range?.from ?? null,
    year_to: selected.year_range?.to ?? null,
    complexity_score: selected.complexity?.score ?? null,
    complexity_level: selected.complexity?.level ?? null,
    iterations: iterations + 1,
    result_count: resultCount,
    patents: fonte === 'ops' && tipo === null ? rawItems : [],
    articles: fonte === 'scopus' && tipo === null ? rawItems : [],
  }
}

export interface SessionProbeQueryRow extends SessionProbeQueryPayload {
  id: number
  session_id: number
  // Documentos persistidos (patent/article) vinculados a essa query, no
  // formato "cru" que o probe search devolveria - só vem populado por GET
  // /research-session/{id} (retomar sessão), usado por sessionHydration.ts.
  documents: Record<string, unknown>[]
}

export interface SessionInputRow {
  id: number
  session_id: number
  parent_id: number | null
  theme: string
  description: string | null
  area_of_study: string | null
  keywords: string[] | null
  year_from: number | null
  year_to: number | null
  iterations: number
}

// Uma chamada de IA (refino de tema, especificação, geração de query
// probe/final) medida no backend e registrada na sessão - log append-only,
// pode ter várias linhas por step (inclusive tentativas descartadas em
// "gerar outras"). Ver schemas/session_input.py:SessionAiCallRow.
export interface SessionAiCallRow {
  id: number
  session_id: number
  step: string
  provider: string
  model: string
  duration_ms: number
  input_tokens: number | null
  output_tokens: number | null
  total_tokens: number | null
  attempts: number
  created_at: string
}

export interface SaveSessionResponse {
  session_id: number
  session_public_id: string
  session_name: string
  completed: boolean
  root: SessionInputRow
  generated: SessionInputRow | null
  probe_queries: SessionProbeQueryRow[]
  ai_calls: SessionAiCallRow[]
}

// Fatia do useFormStore necessária pra montar o payload de save - qualquer
// objeto com pelo menos esses campos serve (ex: `useFormStore.getState()`).
export interface SaveSessionFormState {
  input: FormInput
  step2SelectedTheme: { id: string; theme: string; description: string } | null
  step2Iterations: number
  step3Queries: QueryOptionResult[] | null
  step3SelectedIndex: number | null
  step3Iterations: number
  step3PatentResults: { resultsCount: number; rawItems: Record<string, unknown>[] } | null
  step3ArticleQueries: QueryOptionResult[] | null
  step3ArticleSelectedIndex: number | null
  step3ArticleIterations: number
  step3ArticleResults: { resultsCount: number; rawItems: Record<string, unknown>[] } | null
  // Escolha da Query Final (ver TermSampling.tsx/FinalExploration.tsx) -
  // usados só pra montar a linha de query final (tipo=variante escolhida)
  // quando houver uma. Termos/seleção da Amostragem de Termos em si não são
  // enviados - só a soma das iterações de ambas as etapas.
  step4PatentIterations: number
  step4PatentQueries: Record<FinalQueryVariant, QueryOptionResult> | null
  step4PatentSelectedVariant: FinalQueryVariant | null
  step4PatentQueryIterations: number
  step4PatentResults: { resultsCount: number } | null
  step4ArticleIterations: number
  step4ArticleQueries: Record<FinalQueryVariant, QueryOptionResult> | null
  step4ArticleSelectedVariant: FinalQueryVariant | null
  step4ArticleQueryIterations: number
  step4ArticleResults: { resultsCount: number } | null
  aiCallLog: AiUsage[]
}

export interface SaveSessionPayload {
  root: SessionInputRootPayload
  generated: SessionInputGeneratedPayload | null
  probe_queries: SessionProbeQueryPayload[]
  ai_calls: AiUsage[]
  completed: boolean
}

// Monta o payload de save a partir do estado atual do form store - tolera
// progresso parcial (só o Step1 preenchido já basta: `root.theme` é o único
// campo obrigatório no backend). Usado tanto pelo botão "Salvar progresso"
// (completed=false) quanto pelo botão de finalizar (completed=true).
export function buildSaveSessionPayload(
  formState: SaveSessionFormState,
  completed: boolean,
): SaveSessionPayload {
  const wasRefinedByAI =
    !!formState.step2SelectedTheme && formState.step2SelectedTheme.id !== 'input'
  const generated = wasRefinedByAI
    ? {
        theme: formState.step2SelectedTheme!.theme,
        description: formState.step2SelectedTheme!.description || null,
        iterations: formState.step2Iterations,
      }
    : null

  // `iterations` da linha de probe soma as duas etapas que ainda acontecem
  // dentro da Exploração Inicial: regeneração da própria query probe
  // (Step3, "Gerar outras") + reextração de termos (Amostragem de Termos,
  // "Gerar novos" - substep dessa mesma etapa, sempre roda antes da query
  // final existir).
  const patentQuery = buildProbeQueryPayload(
    formState.step3SelectedIndex !== null ? formState.step3Queries?.[formState.step3SelectedIndex] : undefined,
    'ops',
    formState.step3Iterations + formState.step4PatentIterations,
    formState.step3PatentResults?.resultsCount ?? null,
    formState.step3PatentResults?.rawItems ?? [],
  )
  const articleQuery = buildProbeQueryPayload(
    formState.step3ArticleSelectedIndex !== null
      ? formState.step3ArticleQueries?.[formState.step3ArticleSelectedIndex]
      : undefined,
    'scopus',
    formState.step3ArticleIterations + formState.step4ArticleIterations,
    formState.step3ArticleResults?.resultsCount ?? null,
    formState.step3ArticleResults?.rawItems ?? [],
  )

  // Linha da query final escolhida (Escolha da Query Final), auto-
  // relacionada à linha de probe da mesma fonte - null se o usuário ainda
  // não chegou a escolher uma variante. `iterations` aqui é só a
  // regeneração dessa variante ("Gerar outras" na Escolha da Query Final) -
  // a reextração de termos já foi contabilizada na linha de probe acima.
  const patentFinalQuery = buildProbeQueryPayload(
    formState.step4PatentSelectedVariant !== null
      ? formState.step4PatentQueries?.[formState.step4PatentSelectedVariant]
      : undefined,
    'ops',
    formState.step4PatentQueryIterations,
    formState.step4PatentResults?.resultsCount ?? null,
    [],
    formState.step4PatentSelectedVariant,
  )
  const articleFinalQuery = buildProbeQueryPayload(
    formState.step4ArticleSelectedVariant !== null
      ? formState.step4ArticleQueries?.[formState.step4ArticleSelectedVariant]
      : undefined,
    'scopus',
    formState.step4ArticleQueryIterations,
    formState.step4ArticleResults?.resultsCount ?? null,
    [],
    formState.step4ArticleSelectedVariant,
  )

  return {
    root: mapInputToSessionInputRoot(formState.input),
    generated,
    probe_queries: [patentQuery, articleQuery, patentFinalQuery, articleFinalQuery].filter(
      (q): q is SessionProbeQueryPayload => q !== null
    ),
    ai_calls: formState.aiCallLog,
    completed,
  }
}

// Salva (cria, se `sessionId` for null) ou atualiza (se já existir) a sessão.
// A primeira chamada de "Salvar progresso" ou "Finalizar" cria a sessão via
// POST; qualquer save seguinte (inclusive o finalize de uma sessão retomada)
// faz PUT, evitando duplicar as linhas de session_input/session_probe_query.
export async function saveSession(
  sessionId: number | null,
  name: string,
  payload: SaveSessionPayload,
): Promise<SaveSessionResponse> {
  const body = { name, ...payload }
  const { data } = sessionId
    ? await apiClient.put(`/research-session/${sessionId}`, body)
    : await apiClient.post('/session-input', body)
  return data.data
}

File: schemas/session_input.py
"""
Schemas for saving a prospecting session's input parameters - either partially,
mid-wizard ("save progress", completed=False), or as a full finalization
(completed=True). The same request/response shape backs both the create path
(POST /session-input) and the update path (PUT /research-session/{id}).
"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, field_validator


class SessionInputRoot(BaseModel):
    """Input original do usuário (Step1), raiz da cadeia de session_input."""

    theme: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Tema principal da prospecção tecnológica (obrigatório)",
    )
    description: Optional[str] = Field(default=None, max_length=2000)
    area_of_study: Optional[str] = Field(default=None, max_length=500)
    keywords: Optional[list[str]] = Field(default=None, max_items=50)
    year_from: Optional[int] = Field(default=None)
    year_to: Optional[int] = Field(default=None)

    @field_validator("theme")
    @classmethod
    def validate_theme(cls, value: str) -> str:
        return value.strip()

    @field_validator("description", "area_of_study", mode="before")
    @classmethod
    def validate_optional_text(cls, value: Optional[str]) -> Optional[str]:
        if value is not None:
            stripped = value.strip()
            return stripped if stripped else None
        return None

    @field_validator("keywords", mode="before")
    @classmethod
    def validate_keywords(cls, value: Optional[list[str]]) -> Optional[list[str]]:
        if value is None:
            return None
        normalized = [kw.strip() for kw in value if isinstance(kw, str) and kw.strip()]
        return list(set(normalized)) if normalized else None


class SessionInputGenerated(BaseModel):
    """Variante gerada/refinada por IA escolhida para seguir adiante."""

    theme: str = Field(..., min_length=1, max_length=500)
    description: Optional[str] = Field(default=None, max_length=2000)
    iterations: int = Field(default=0, ge=0)


class TermInput(BaseModel):
    """
    Termo extraído por NLP local (spaCy + KeyBERT + TF-IDF, não é IA) na Amostragem de Termos, a partir dos
    documentos de uma probe query, com a flag `selected` indicando se o
    usuário marcou esse termo pra usar na construção da query final. Ver
    ProbeQueryTerm.
    """

    term: str = Field(..., min_length=1, max_length=255)
    score: float
    frequency: int = Field(default=0, ge=0)
    selected: bool = Field(default=False)


class SessionProbeQueryInput(BaseModel):
    """
    Query do Step3 (patente ou artigo) selecionada pelo usuário, ou a query
    final escolhida na Amostragem de Termos/Escolha da Query Final -
    diferenciadas por `tipo` (None = probe, "specific"|"balanced"|"generic"
    = a variante final escolhida). Ver SessionProbeQuery.
    """

    fonte: str = Field(..., pattern="^(ops|scopus)$")
    tipo: Optional[str] = Field(default=None, pattern="^(specific|balanced|generic)$")
    query_text: str = Field(..., min_length=1)
    fields: Optional[dict[str, list[str]]] = None
    year_from: Optional[int] = None
    year_to: Optional[int] = None
    complexity_score: Optional[float] = None
    complexity_level: Optional[str] = None
    iterations: int = Field(default=1, ge=1)
    result_count: Optional[int] = Field(default=None, ge=0)
    patents: list[dict[str, Any]] = Field(default_factory=list, max_items=200)
    articles: list[dict[str, Any]] = Field(default_factory=list, max_items=200)
    # Termos da Amostragem de Termos (todos os extraídos, com `selected`
    # marcando os escolhidos) - só preenchido em linhas de probe (tipo=None).
    terms: list[TermInput] = Field(default_factory=list, max_items=100)


class SessionAiCallInput(BaseModel):
    """Uma chamada de IA (refino de tema, especificação, geração de query
    probe/final) medida no backend e reenviada pelo frontend junto do save -
    sempre acrescentada como linha nova (log), nunca upsertada."""

    step: str = Field(..., max_length=50)
    provider: str = Field(..., max_length=50)
    model: str = Field(..., max_length=100)
    duration_ms: float = Field(..., ge=0)
    input_tokens: Optional[int] = Field(default=None, ge=0)
    output_tokens: Optional[int] = Field(default=None, ge=0)
    total_tokens: Optional[int] = Field(default=None, ge=0)
    attempts: int = Field(default=1, ge=1)


class SessionAiCallRow(SessionAiCallInput):
    """Representação persistida de uma linha session_ai_call."""

    id: int
    session_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class SessionProbeQueryRow(SessionProbeQueryInput):
    """Representação persistida de uma linha session_probe_query."""

    id: int
    session_id: int
    result_count: Optional[int] = None
    # Documentos persistidos (patent/article) vinculados a essa query,
    # reconstruídos no formato "cru" que o probe search devolveria (ver
    # patent_to_raw_item/article_to_raw_item) - só populado por GET
    # /research-session/{id} (retomar sessão), não faz parte do payload de save.
    documents: list[dict[str, Any]] = Field(default_factory=list)

    class Config:
        from_attributes = True


class SessionInputSaveRequest(BaseModel):
    """Payload enviado ao salvar/atualizar uma sessão: nome + input raiz +
    gerado (opcional) + queries de probe (opcionais) + se a sessão deve ficar
    marcada como concluída ou apenas salva em progresso."""

    name: str = Field(..., min_length=1, max_length=255)
    root: SessionInputRoot
    generated: Optional[SessionInputGenerated] = None
    # Até 4: 1 linha de probe + 1 de query final, por fonte (ops/scopus).
    probe_queries: list[SessionProbeQueryInput] = Field(default_factory=list, max_items=4)
    ai_calls: list[SessionAiCallInput] = Field(default_factory=list)
    completed: bool = Field(default=False)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        return value.strip()


class SessionInputRow(BaseModel):
    """Representação persistida de uma linha session_input."""

    id: int
    session_id: int
    parent_id: Optional[int] = None
    theme: str
    description: Optional[str] = None
    area_of_study: Optional[str] = None
    keywords: Optional[list[str]] = None
    year_from: Optional[int] = None
    year_to: Optional[int] = None
    iterations: int

    class Config:
        from_attributes = True


class SessionInputSaveResponse(BaseModel):
    """Resultado do save/update: sessão + linhas de input persistidas."""

    session_id: int
    session_public_id: str
    session_name: str
    completed: bool
    root: SessionInputRow
    generated: Optional[SessionInputRow] = None
    probe_queries: list[SessionProbeQueryRow] = Field(default_factory=list)
    ai_calls: list[SessionAiCallRow] = Field(default_factory=list)
